---
title: "rbp_classifier"
output: html_document
---

## RBP classifier 

This work builds a model to classify RNA binding proteins (RBPs) based off their CTD data.

The data used to train the model is available here: 

http://rnabinding.com/RBPPred.html
http://rnabinding.com/software.html


### Packages

```{r}

library(protr)
library(tidyverse)
library(annotate)
library(org.Hs.eg.db)
library(randomForest)

```

### Set-up 

Download and read in data:

```{bash}

wget -O RBP2780.fasta http://rnabinding.com/RBP2780 &
wget -O non-RBP7093.fasta http://rnabinding.com/non-RBP7093 &

```

```{r}

RBPs <- readFASTA("RBP2780.fasta")
non_RBPs <- readFASTA("non-RBP7093.fasta")

```



Process the data -- create dataframes for RNA binding proteins and non-RNA binding proteins, annotate them, then join them together: 

```{r}

RBP_df <- as.data.frame(do.call(rbind, RBPs)) 
rownames(RBP_df) <- sapply(strsplit(rownames(RBP_df),"[|]"), `[`, 2)
colnames(RBP_df) <- "Sequence"
RBP_df["Protein_type"] <- "RBP"

```

```{r}

cols <- c("ENTREZID","SYMBOL")
annotation <- AnnotationDbi::select(org.Hs.eg.db, keys=rownames(RBP_df), columns=cols, keytype="UNIPROT")

```

```{r}

non_RBP_df <- as.data.frame(do.call(rbind, non_RBPs)) 
colnames(non_RBP_df) <- "Sequence"
non_RBP_df["Protein_type"] <- "Non-RBP"

```

```{r}

RBP_df <- rbind(RBP_df, non_RBP_df)

```


Calculate CTD descriptors per protein: 

```{r}

df <- sapply(RBP_df[,1], extractCTDC)

```

'Error in FUN(X[[i]], ...) : x has unrecognized amino acid type'


Function failed due to non-standard amino acid types. Check whether there are non-standard residues in each sequence: 

```{r}

residue_check <- function(x) {
    AADict <- c("A", "R", "N", "D", "C", 
        "E", "Q", "G", "H", "I", 
        "L", "K", "M", "F", "P", 
        "S", "T", "W", "Y", "V")
    all(strsplit(x, split = "")[[1]] %in% AADict)
}


```

Create index of sequences with non-standard residues: 

```{r}
non_standard_residues <- list()

for (i in 1:nrow(RBP_df)){
  non_standard_residues[i] <- residue_check(RBP_df[i,1])
}

non_standard_residues <- as.data.frame(do.call(rbind, non_standard_residues)) 

```

Subset RBP data by index: 

```{r}

RBP_df_clean <- RBP_df[non_standard_residues$V1,]

```


Recalculate CTD descriptors per sequence and append to dataframe: 

```{r}

df <- t(sapply(RBP_df_clean[,1], extractCTDC))

```

```{r}

RBP_df_clean <- cbind(RBP_df_clean, df)

```


### Random forest

Set-up with default parameters: 

```{r}

labels <- RBP_df_clean$Protein_type
data <- RBP_df_clean[, 3:ncol(RBP_df_clean)]

```

```{r}

set.seed(111)

indices <- c(sample(which(labels=="RBP"), .8*length(which(labels=="RBP"))),
             sample(which(labels=="Non-RBP"), .8*length(which(labels=="Non-RBP"))))

data_train <- data[indices, ]
labels_train <- labels[indices]
data_test <- data[-indices, ]
labels_test <- labels[-indices]

```


```{r}

set.seed(111)
rfModel <- randomForest(x = data_train, y = factor(labels_train), proximity=TRUE)

```

```{r}

predictions <- predict(rfModel, newdata = data_test)

```

```{r}

table(labels_test, predictions)

```

```{r}

prediction_prob <- predict(rfModel, newdata = data_test, type = "prob")

```

```{r}

library(ROCR)
pred_object <- prediction(prediction_prob[, 2], labels_test)
perf_object <- performance(pred_object, "tpr", "fpr")
plot(perf_object, colorize = T)

```

```{r}

AUC <- performance(pred_object, measure = "auc")@y.values
print(AUC)

```

AUC = 0.8773491
Pretty good, could possibly be better... 

Next, measure importance of variables: 

```{r}

var_imp <- rfModel$importance
var_imp <- data.frame(Variable = rownames(var_imp), MeanDecreaseGini = var_imp[,
1], stringsAsFactors = F)
var_imp <- var_imp[order(var_imp$MeanDecreaseGini, decreasing = T), ]

```


```{r}
ggplot(data = var_imp[1:20, ], aes(x = reorder(Variable, MeanDecreaseGini),
y = MeanDecreaseGini)) + geom_bar(stat = "identity", fill = "steelblue") +
coord_flip() + theme_bw() + ylab("Mean decrease in Gini index") + xlab("Variable") +
ggtitle("Variable importance") + theme(plot.title = element_text(hjust = 0.5))

```

Charge.group1 and polarity.group1 are the most important by a large margain, then followed by secondarystruct.group2 and hydrophobocity.group 3. 

Could the model be improved? Investigate different values of ntree & mtry.

Firstly use out-of-bag (OOB) error to probe how number of trees effects accuracy... 

```{r}

oob_error_data <- data.frame(
  Trees=rep(1:nrow(rfModel$err.rate), times=3),
  Type=rep(c("OOB", "Non-RBP", "RBP"), each=nrow(rfModel$err.rate)),
  Error=c(rfModel$err.rate[,"OOB"], 
    rfModel$err.rate[,"Non-RBP"], 
    rfModel$err.rate[,"RBP"]))


```


```{r}

ggplot(data=oob_error_data, aes(x=Trees, y=Error)) +
  geom_line(aes(color=Type))

ggsave("oob_error_rate_500_trees.pdf")

```

The error starts to plateau after 100 trees and has levelled out by 200. 

Now optimise for mtry... 

```{r}

oob_values <- vector(length=10)
AUC_values <- vector(length=10)


for(i in 1:10) {
  temp.model <- randomForest(x = data_train, y = factor(labels_train), mtry=i)
  oob_values[i] <- temp.model$err.rate[nrow(temp.model$err.rate),1]
  
  prediction_prob <- predict(temp.model, newdata = data_test, type = "prob")
  pred_object <- prediction(prediction_prob[, 2], labels_test)
  AUC_values[i] <- performance(pred_object, measure = "auc")@y.values[[1]]
}

oob_values
AUC_values

```

```{r}

min(oob_values)
max(AUC_values)

which(oob_values == min(oob_values))
which(AUC_values == max(AUC_values))
```

Not much difference overall but mtry = 4 gives best AUC value overall so use that in final model.

```{r}

set.seed(111)
rfModel <- randomForest(x = data_train,
                        y = factor(labels_train),
                        proximity=TRUE,
                        mtry=which(AUC_values == max(AUC_values)))

predictions <- predict(rfModel, newdata = data_test)
prediction_prob <- predict(rfModel, newdata = data_test, type = "prob")

pred_object <- prediction(prediction_prob[, 2], labels_test)
perf_object <- performance(pred_object, "tpr", "fpr")
plot(perf_object, colorize = T)

```

```{r}

AUC <- performance(pred_object, measure = "auc")@y.values
print(AUC)

```


```{r}

distance.matrix <- as.dist(1-rfModel$proximity)

mds.stuff <- cmdscale(distance.matrix, eig=TRUE, x.ret=TRUE)

# calculate the percentage of variation that each MDS axis accounts for...
mds.var.per <- round(mds.stuff$eig/sum(mds.stuff$eig)*100, 1)

# plot MDS axes and the variation:
mds.values <- mds.stuff$points
mds.data <- data.frame(Sample=rownames(mds.values),
  X=mds.values[,1],
  Y=mds.values[,2],
  Status=labels_train)

ggplot(data=mds.data, aes(x=X, y=Y, label=Sample)) + 
  geom_text(aes(color=Status)) +
  theme_bw() +
  xlab(paste("MDS1 - ", mds.var.per[1], "%", sep="")) +
  ylab(paste("MDS2 - ", mds.var.per[2], "%", sep="")) +
  ggtitle("MDS plot using (1 - Random Forest Proximities)")

```


### Use model to predict RBP in human proteome 

Have established the model is quite conservative when it comes to making predictions so unlikely to encounter false positives. 

Read in the whole human proteome: 

```{r}

proteome <- readFASTA("uniprot-proteome UP000005640.fasta")

```

Process dataframe: 

```{r}

proteome_df <- as.data.frame(do.call(rbind, proteome)) 
rownames(proteome_df) <- sapply(strsplit(rownames(proteome_df),"[|]"), `[`, 2)
colnames(proteome_df) <- "Sequence"
proteome_df["Protein_type"] <- "Unknown"

```

Remove non-standard residues: 

```{r}
non_standard_residues <- list()

for (i in 1:nrow(proteome_df)){
  non_standard_residues[i] <- residue_check(proteome_df[i,1])
}

non_standard_residues <- as.data.frame(do.call(rbind, non_standard_residues)) 
proteome_df_clean <- proteome_df[non_standard_residues$V1,]

```


Calculate CTD descriptors for whole proteome: 

```{r}

df <- t(sapply(proteome_df_clean[,1], extractCTDC))
proteome_df_clean <- cbind(proteome_df_clean, df)

```

Make predictions:

```{r}

predictions <- predict(rfModel, newdata = proteome_df_clean[, 3:ncol(proteome_df_clean)])
prediction_prob <- as.data.frame(predict(rfModel, newdata = data_test, type = "prob"))

```

```{r}

predicted_RBPS <- prediction_prob[prediction_prob$RBP > 0.6,]

```

```{r}

cols <- c("SYMBOL", "ENTREZID")
annotation <- AnnotationDbi::select(org.Hs.eg.db, keys=rownames(predicted_RBPS), columns=cols, keytype="UNIPROT")

predicted_RBPS <- cbind(predicted_RBPS , annotation)

```

Manually inspected several of the top candidate RBPs and found evidence they were RBPs.
To take this work further could extract GO terms for each of the candidate RBPs and see if 'RNA binding' is among them. 
Then take the remainder and see what information is available for them, if poorly annotated this could be a way to uncover candidate RBPS. 

Worth noting the model is very conservative, likely missing several potential RBPs. Could try different classifiers and see what effect this has on the results. 


```{r}

write.csv(predicted_RBPS, "predicted_RBPS.csv")

```


