# Analyis of Fleischer alternative polyadenylation data 

Load libraries

```{r}

library(tidyverse)
library(factoextra)
library(reshape)
library(pvca)
library(Biobase)
library(glmnet)

```

## Initial processing

Read in data:

```{r}

data <- read.csv("fleischer_sep_roar_results.csv", header = TRUE)
data <- as.matrix(data)

```


Read in meta-data data:

```{r}

metadata <- read.csv("fleischer_metadata.csv", header = TRUE)

```

Format data for preliminary analysis. First extract only columns with data.
Then convert matrix into a numeric type, before converting to dataframe. 
Add additional column with gene names.


```{r}

data_plain <- data[,c(1:136,139:274)]
mode(data_plain) = "numeric"
data_plain <- data.frame(data_plain)
data_plain$gene_ID <- data[,138]

```

Remove rows corresponding to NA values for gene names.
Then use gene names for rownames. 

```{r}

data_plain <- data_plain[!duplicated(data_plain$gene_ID), ]
data_plain <- data_plain[!is.na(data_plain$gene_ID), ]
rownames(data_plain) <- data_plain$gene_ID

```

Seperate the TPM data and ratio data. Filter the TPM data to exclude counts below 1. 
Use the newly fitlered TPM data to subset the ratio data. 
Then further filter that ratio data to remove outliers. 

```{r}

data_plain_TPM <- data_plain[,137:272]


for(i in 1:ncol(data_plain_TPM)){
    data_plain_TPM <- data_plain_TPM %>% filter(., .[i] > 1)
   
}



data_plain_ratio <- data_plain[,1:136]
data_plain_ratio <- data_plain_ratio[rownames(data_plain_ratio) %in% rownames(data_plain_TPM),]



for(i in 1:ncol(data_plain_ratio)){
    data_plain_ratio <- data_plain_ratio %>% filter(., .[i] > 0 & .[i] <20) 
   
}


```


Process the metadata file for use in analysis. Firstly, convert the HPS ages into numeric values, then conver the whole column into numeric values using 'as.numeric'. 

Then extract only the columns of interest for the analysis. Add an additional column to divide the samples into age groups for visualisation purposes. Make sure the HPS samples have their own unique identifier. 

```{r}
metadata_original <- metadata

metadata[127,2] <- 8
metadata[128,2] <- 8
metadata[129,2] <- 2.25
metadata[130,2] <- 3.75
metadata[131,2] <- 4.75
metadata[132,2] <- 8.5
metadata[133,2] <- 7
metadata[134,2] <- 5
metadata[135,2] <- 8.75
metadata[136,2] <- 3
metadata$age <- as.numeric(metadata$age)

metadata <- metadata %>% dplyr::select("Run", "age", "sex", "disease", "source_name", "ethnicity", "Instrument") %>% mutate(
    Age_group = case_when(
    age >= 60 ~ "Old",
    age > 30 & age < 60 ~ "Middle-aged",
    age <= 30 ~ "Young"
    ))

rownames(metadata) <- metadata$Run
metadata[127:136,8] <- "HPS"

```



Look at boxplots/distribution: 

```{r}

boxplot(data_plain_ratio,ylab="Ratio",xlab="Sample")

```

```{r}

plot(density(data_plain_ratio[,1]), ylim=c(0,0.7), xlim=c(-1,20))

for(i in 2:ncol(data_plain_ratio)){ 
    lines(density(data_plain_ratio[,i])) 

}

```

```{r}

boxplot(log2(data_plain_TPM),ylab="TPM",xlab="Sample")

```

## Ratio results: 

### Exploratory analysis 

```{r}

pca_res <- prcomp(t(data_plain_ratio))

df_out <- as.data.frame(pca_res$x)
p<-ggplot(df_out,aes(x=PC1,y=PC2,color=metadata$Age_group ))
p<-p+geom_point()
p

```

Calculate percentage of variance each PC contributes: 

```{r}

# Manually calculating 
pca.var <- pca_res$sdev^2
pca.var.per <- round(pca.var/sum(pca.var)*100, 1)

barplot(pca.var.per[1:20],
        main="Scree Plot - Top20 PCs",
        xlab="Principal Components",
        ylab="Percent Variation",
        names.arg = pca.var.per[1:20])

```

There is a package that makes exploratory analysis of multivariate data much more rapid: 

https://cran.r-project.org/web/packages/factoextra/readme/README.html

Using factoextra instead:

```{r}

fviz_eig(pca_res)

```

```{r}

fviz_pca_biplot(pca_res, repel = TRUE,
                col.var = "#2E9FDF", # Variables color
                col.ind = "#696969", # Individuals color
                )

```

```{r}

# Contributions of variables to PC1
fviz_contrib(pca_res, choice = "var", axes = 1, top = 30)

# Contributions of variables to PC2
fviz_contrib(pca_res, choice = "var", axes = 2, top = 30)

```

Of interest: GORAB, STRAP, YAF2 SMAD4, ENAH
Several are ageing related or ER proteins

```{r}

genes <- cbind.data.frame(t(data_plain_ratio["GORAB",1:126]),
               t(data_plain_ratio["STRAP",1:126]),
               t(data_plain_ratio["YAF2",1:126]),
               t(data_plain_ratio["SMAD4",1:126]),
               t(data_plain_ratio["ENAH",1:126]),
               metadata$age[1:126])

genes <- melt(genes, id.vars = "metadata$age[1:126]")


colnames(genes) <- c("Age", "Gene Name", "Ratio short:long isoform")


ggplot(genes, aes(x=Age, y=`Ratio short:long isoform`, color=`Gene Name`)) +
    geom_point() + 
    geom_smooth(method=lm, se=FALSE)

```

GORAB is especially interesting because the HPS samples are more similar to the middle-aged/old samples.
In geroderma osteodysplastica GORAB is mutated and skin becomes loose and wrinkly. Short isoforms of of mRNA generally escape regulation by miRNAs - could it be HPS patients express longer isoforms of GORAB? This might explain some of the similarities in phenotypes.

Additionally, GORAB is a golgi apparatus protein of unknown function - could it be affecting collagen synthesis/regulation? 

Publication: 

http://europepmc.org/article/MED/30631079

Current grant to elucidate function: 

https://gtr.ukri.org/projects?ref=MR%2FN000366%2F1


```{r}

GORAB <- cbind.data.frame(t(data_plain_ratio["GORAB",]),
               metadata$age,
               metadata$Age_group)

colnames(GORAB) <- c("GORAB short:long isoform ratio", "Age", "Age group")


ggplot(GORAB,
       aes(x=Age, y=`GORAB short:long isoform ratio`, color=`Age group`)) +
    geom_point() 

```


PVCA estimation of batch effects: 

(Original code provided by Liverpool's CBF)

```{r}

pheno_formatted <- new("AnnotatedDataFrame", data = metadata[,2:7])
ratio_data_formatted <- as.matrix(data_plain_ratio)
colnames(ratio_data_formatted) <- gsub(pattern = "_.*",
                                       replacement = "",
                                       x = colnames(ratio_data_formatted))


PVCA_data <- ExpressionSet(assayData = ratio_data_formatted, phenoData = pheno_formatted)


pvcaObj <- pvcaBatchAssess(abatch = PVCA_data,
                           batch.factors = colnames(pheno_formatted@data),
                           threshold = 0.7)

pvca_res <- data.frame(as.data.frame(pvcaObj$label),
                       t(as.data.frame(pvcaObj$dat)))

colnames(pvca_res) <- c("effect", "variance")
pvca_res <- pvca_res[-nrow(pvca_res), ]


ggplot((pvca_res), aes(x= effect, y = variance)) +
    geom_bar(stat = 'identity', position = 'dodge', col ='transparent') +
    scale_fill_discrete(guide = 'none') +
    theme_bw(base_size = 16) +
    theme(plot.title = element_text(hjust=0.5),
          axis.text.x=element_text(angle=45,hjust=1)) +
    labs(x = 'Effects', y = 'Weighted average \n proportion variance') +
    ggtitle("PVCA estimation bar chart-non corrected data")

```

Two different instruments were used for sequencing and contribute a batch effect. Try to remove batches before doing additional modelling via glmnet or MOFA.



## TPM results: 


```{r}

pca_res <- prcomp(t(data_plain_TPM), scale. = TRUE)

df_out <- as.data.frame(pca_res$x)
p<-ggplot(df_out,aes(x=PC1,y=PC2,color=metadata$Age_group ))
p<-p+geom_point()
p


```

```{r}

fviz_eig(pca_res)

```

```{r}

# Contributions of variables to PC1
PC1 <- fviz_contrib(pca_res, choice = "var", axes = 1, top = 30)
PC1

# Contributions of variables to PC2
PC2 <- fviz_contrib(pca_res, choice = "var", axes = 2, top = 30)
PC2

# Contributions of variables to PC3
PC3 <- fviz_contrib(pca_res, choice = "var", axes = 3, top = 30)
PC3

```

```{r}

genes <- cbind.data.frame(t(data_plain_TPM["SSNA1",1:126]),
               t(data_plain_TPM["PPP1R14B",1:126]),
               t(data_plain_TPM["MRPS12",1:126]),
               t(data_plain_TPM["PPP1CA",1:126]),
               t(data_plain_TPM["PDCD4",1:126]),
               metadata$age[1:126])

genes <- melt(genes, id.vars = "metadata$age[1:126]")


colnames(genes) <- c("Age", "Gene Name", "TPM")


ggplot(genes, aes(x=Age, y=TPM, color=`Gene Name`)) + 
    geom_point() + 
    geom_smooth(method=lm)

```

Most interesting are PPP1R14B and PDCD4 (greatest slopes). 

https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0750-x
https://www.researchgate.net/publication/306067867_PDCD4_is_a_CSL_associated_protein_with_a_transcription_repressive_function_in_cancer_associated_fibroblast_activation


Next extract collagen specific data:

```{r}

genes <- cbind.data.frame(t(data_plain_ratio["COL1A1",1:126]),
               t(data_plain_ratio["COL1A2",1:126]),
               metadata$age[1:126])

genes <- melt(genes, id.vars = "metadata$age[1:126]")


colnames(genes) <- c("Age", "Gene Name", "Ratio short:long isoform")


ggplot(genes, aes(x=Age, y=`Ratio short:long isoform`, color=`Gene Name`)) + 
    geom_point() + 
    geom_smooth(method=lm)

```


```{r}

COL1A1 <- data_plain["COL1A1",-ncol(data_plain)]
COL1A2 <- data_plain["COL1A2",-ncol(data_plain)]

col_df <- t(rbind(COL1A1, COL1A2))

ratio <- as.data.frame(col_df[grepl("ratio", rownames(col_df)), ])
TPM <- as.data.frame(col_df[grepl("TPM", rownames(col_df)), ])

rownames(TPM) <-  gsub(pattern = "_.*",
                                    replacement = "",
                                    x = rownames(TPM))

rownames(ratio) <- gsub(pattern = "_.*",
                                    replacement = "",
                                    x = rownames(ratio))

col_df <- cbind(TPM, ratio)
colnames(col_df) <- c("COL1A1_TPM", "COL1A2_TPM", "COL1A1_ratio", "COL1A2_ratio")


```


```{r}

col_df <- col_df %>% mutate(
    COL1A1_short_form = COL1A1_TPM/(COL1A1_ratio+1)*COL1A1_ratio,
    COL1A1_long_form = COL1A1_TPM/(COL1A1_ratio+1),
    COL1A2_short_form = COL1A2_TPM/(COL1A2_ratio+1)*COL1A2_ratio,
    COL1A2_long_form = COL1A2_TPM/(COL1A2_ratio+1)) %>% add_column(metadata$Age_group, metadata$sex)

# (round(test$COL1A2_long_form) + round(test$COL1A2_short_form)) == round(test$COL1A2_TPM)

```


```{r}

write.csv(col_df, "./fleischer_collagen_data.csv")

```



